vortex-serde/src/layouts/read/layouts/chunked.rs

```diff
@@ -81,33 +83,267 @@ impl ChunkedLayout {
             .map(|b| b.bytes()[0] != 0)
             .unwrap_or(false)
     }
+
+    fn children(&self) -> impl Iterator<Item = (usize, footer::Layout)> {
+        self.flatbuffer()
+            .children()
+            .unwrap_or_default()
+            .iter()
+            .enumerate()
+            .skip(if self.has_metadata() { 1 } else { 0 })
+    }
+
+    fn child_ranges(&self) -> Vec<(usize, usize)> {
+        self.children()
+            .map(|(_, c)| c.length())
+            .scan(0u64, |acc, length| {
+                let current = *acc;
+                *acc += length;
+                Some((current as usize, *acc as usize))
+            })
+            .collect::<Vec<_>>()
+    }
+
+    fn child_layouts<C: Fn(LayoutPartId) -> RelativeLayoutCache>(
+        &self,
+        cache: C,
+    ) -> VortexResult<VecDeque<RangedLayoutReader>> {
+        self.children()
+            .zip_eq(self.child_ranges())
+            .map(|((i, c), (begin, end))| {
+                let layout = self.layout_builder.read_layout(
+                    self.fb_bytes.clone(),
+                    c._tab.loc(),
+                    self.scan.clone(),
+                    cache(i as u16),
+                )?;
+                Ok(((begin, end), layout))
+            })
+            .collect::<VortexResult<VecDeque<_>>>()
+    }
 }
 
 impl LayoutReader for ChunkedLayout {
-    fn read_next(&mut self) -> VortexResult<Option<ReadResult>> {
-        if let Some(cr) = &mut self.reader {
-            cr.read()
+    fn add_splits(&self, row_offset: usize, splits: &mut BTreeSet<usize>) -> VortexResult<()> {
+        for ((begin, _), child) in self.child_layouts(|i| self.message_cache.unknown_dtype(i))? {
+            child.add_splits(row_offset + begin, splits)?
+        }
+        Ok(())
+    }
+
+    fn read_selection(&mut self, selector: RowMask) -> VortexResult<Option<BatchRead>> {
+        if let Some(br) = &mut self.chunk_reader {
+            br.read_next(selector)
         } else {
-            let children = self
-                .flatbuffer()
-                .children()
-                .ok_or_else(|| vortex_err!("Missing children"))?
-                .iter()
-                .enumerate()
-                // Skip over the metadata table of this layout
-                .skip(if self.has_metadata() { 1 } else { 0 })
-                .map(|(i, c)| {
-                    self.layout_builder.read_layout(
-                        self.fb_bytes.clone(),
-                        c._tab.loc(),
-                        self.scan.clone(),
-                        self.message_cache
-                            .relative(i as u16, self.message_cache.dtype().clone()),
-                    )
-                })
-                .collect::<VortexResult<VecDeque<_>>>()?;
-            self.reader = Some(BufferedReader::new(children, self.scan.batch_size));
-            self.read_next()
+            self.chunk_reader = Some(BufferedLayoutReader::new(self.child_layouts(|i| {
```

Comment:
Why is this lazily constructed?

Reply:
I wanted all children to be lazily constructed in case of pruning